File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/items/medicine/MedicalItem.java
package de.limbusdev.guardianmonsters.guardians.items.medicine;

import com.badlogic.gdx.math.MathUtils;

import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian;

/**
 * Medicine
 *
 * @author Georg Eckert 2017
 */

public class MedicalItem extends AMedicalItem
{
    public enum Type
    {
        REVIVE, HP_CURE, MP_CURE, STATUS_CURE,
    }

    private int value;
    private Type type;

    /**
     * Implementation of @link{AMedicalItem}, do not use this outside of this module, always use
     * AMedicalItem.
     * @param name
     * @param value
     * @param type
     */
    public MedicalItem(String name, int value, Type type)
    {
        super(name);
        this.value = value;
        this.type = type;
    }

    @Override
    public void apply(AGuardian m)
    {
        switch(type)
        {
            case REVIVE:
                m.getIndividualStatistics().healHP(MathUtils.round(m.getIndividualStatistics().getHPmax()*value/100f));
                break;
            case HP_CURE:
                m.getIndividualStatistics().healHP(value);
                break;
            case MP_CURE:
                m.getIndividualStatistics().healMP(value);
                break;
            default:
                break;
        }
    }

    @Override
    public boolean applicable(AGuardian m)
    {
        switch(type) {
            case REVIVE:
                return m.getIndividualStatistics().isKO();
            case HP_CURE:
                return (m.getIndividualStatistics().isFit() && m.getIndividualStatistics().getHP() < m.getIndividualStatistics().getHPmax());
            case MP_CURE:
                return m.getIndividualStatistics().getMP() < m.getIndividualStatistics().getMPmax();
            default:
                return false;
        }
    }
}

File: GuardianMonsters/guardians/src/de/limbusdev/guardianmonsters/guardians/items/medicine/MedicalItem.kt
package de.limbusdev.guardianmonsters.guardians.items.medicine

import com.badlogic.gdx.math.MathUtils

import de.limbusdev.guardianmonsters.guardians.monsters.AGuardian

/**
 * Medicine
 *
 * @author Georg Eckert 2017
 */

/**
 * Implementation of @link{AMedicalItem}, do not use this outside of this module, always use
 * AMedicalItem.
 * @param name
 * @param value
 * @param type
 */
class MedicalItem
(
        name                : String = "bread",
        val value           : Int = 100,
        override val type   : Type = Type.HP_CURE
)
    : AMedicalItem(name)
{
    override fun apply(guardian: AGuardian)
    {
        when (type)
        {
            Type.REVIVE  -> guardian.stats.healHP(MathUtils.round(guardian.stats.hpMax * value / 100f))
            Type.HP_CURE -> guardian.stats.healHP(value)
            Type.MP_CURE -> guardian.stats.healMP(value)
            else -> {}
        }
    }

    override fun applicable(guardian: AGuardian): Boolean
    {
        return when (type)
        {
            Type.REVIVE  -> guardian.stats.isKO
            Type.HP_CURE -> guardian.stats.isFit && guardian.stats.hp < guardian.stats.hpMax
            Type.MP_CURE -> guardian.stats.mp < guardian.stats.mPmax
            else         -> false
        }
    }
}